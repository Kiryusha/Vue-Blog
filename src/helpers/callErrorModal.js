// calling error modal
export default function (context, error) {
  let errText = 'Неизвестная ошибка.';

  if (error.response) {
    errText = error.response.data.message ? error.response.data.message :
      error.response.data;
  } else {
    errText = error.message;
  }

  context.$Progress.fail();
  context.$modal.show('response', { message: errText });
}
