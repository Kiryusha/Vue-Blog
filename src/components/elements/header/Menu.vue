<template lang="pug">
  .menu
    .item
      router-link.link(
        :to="'/blog/'"
      ) Блог
    .item(v-if="userId")
      router-link.link(
        :to="'/publish/'"
      ) Опубликовать
    .item(v-else)
      span.link(
        @click="$modal.show('auth')"
      ) Войти
    .item(v-if="userId")
      span.link(
        @click="logout()"
      ) Выйти
</template>

<script>
import { mapState, mapActions } from 'vuex';
import callErrorModal from '@/helpers/callErrorModal';

export default {
  data() {
    return {
      response: null,
    };
  },
  computed: {
    ...mapState({
      userId: state => state.user.id,
    }),
  },
  methods: {
    ...mapActions(['authLogout']),
    async logout() {
      try {
        this.$Progress.start();

        await this.authLogout();

        this.$Progress.finish();

        if (this.$route.path.lastIndexOf('/publish/', 0) === 0) {
          this.$router.push('/blog/');
        }
      } catch (error) {
        callErrorModal(this, error);
      }
    },
  },
};
</script>

<style lang="stylus" scoped>
  .menu
    display flex
    padding-top 5px
    flex-wrap wrap
    font-family 'Roboto Slab', sans-serif

  .item
    font-size 14px
    position relative
    padding-right 6px

    &:after
      content '/'
      position absolute
      top 0
      right 0

    &:first-child
      padding-left 6px

      &:before
        content '/'
        position absolute
        top 0
        left 0

  .link
    a-reset()
    margin 0 16px
    cursor pointer

    &._active
      font-weight bold

  +phone()
    flex-direction column

    .item,
    .item:first-child
      padding 0
      margin-top 10px

      &:after,
      &:before
        display none

    .link,
    .item:first-child .link
      font-size 18px
      color #f

</style>
