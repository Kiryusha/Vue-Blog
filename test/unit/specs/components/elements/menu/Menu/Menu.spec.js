// Menu.vue

import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import Menu from 'Components/elements/header/Menu.vue';
import callErrorModal from '@/helpers/callErrorModal';

const localVue = createLocalVue();
localVue.use(Vuex);

jest.mock('@/helpers/callErrorModal', () => jest.fn());

describe('Menu.vue', () => {
  let config = {
    localVue,
    stubs: ['router-link'],
    mocks: {
      $Progress: {
        start: () => {},
        finish: () => {},
      },
      $route: {
        path: '/blog/',
      },
      $router: {
        push: jest.fn(),
      },
      $modal: {
        show: () => {},
      },
    },
  };

  let storeConfig = {
    state: {
      user: {
        id: '1a',
      },
    },
    actions: {
      authLogout: () => {},
    },
  };
  let store = new Vuex.Store(storeConfig);

  describe('component shallow rendering', () => {
    it('should match a snapshot', () => {
      const wrapper = shallowMount(Menu, { ...config, store });

      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('custom actions', () => {
    it('should call authLogout() on logout()', async () => {
      const authLogout = jest.fn().mockResolvedValue();
      const wrapper = shallowMount(Menu, { ...config, store, methods: {
        authLogout,
      }});

      await wrapper.vm.logout()

      expect(authLogout).toHaveBeenCalledTimes(1);
    });

    it('should call callErrorModal() on logout() with error', async () => {
      const authLogout = jest.fn().mockRejectedValue();
      const wrapper = shallowMount(Menu, { ...config, store, methods: {
        authLogout,
      }});

      await wrapper.vm.logout()

      expect(callErrorModal).toHaveBeenCalledTimes(1);
    });

    it('should redirect to "/blog/" on logout() if route is "/publish/"', async () => {
      config.mocks.$route.path = '/publish/';

      const authLogout = jest.fn().mockResolvedValue();
      const wrapper = shallowMount(Menu, { ...config, store, methods: {
        authLogout,
      }});

      await wrapper.vm.logout()

      expect(wrapper.vm.$router.push).toHaveBeenCalledTimes(1);
      expect(wrapper.vm.$router.push).toHaveBeenCalledWith('/blog/');
    });
  });
});
