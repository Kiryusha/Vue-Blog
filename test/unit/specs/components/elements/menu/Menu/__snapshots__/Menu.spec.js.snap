// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu.vue component shallow rendering should match a snapshot 1`] = `
<div class="menu">
  <div class="item">
    <router-link-stub to="/blog/" class="link"></router-link-stub>
  </div>
  <div class="item">
    <router-link-stub to="/publish/" class="link"></router-link-stub>
  </div>
  <div class="item"><span class="link">Выйти</span></div>
</div>
`;
